[feat] Ringseries: added rs_cosh_sinh to caluclate the faster cosh, sinh series together.

Jatinbhardwaj-093 opened the pull request:
### Summary of changes

1. Fixed a typo in the puiseux documentation.
2. TODO of `rs_series_from_list`.
3. Added `rs_cosh_sinh` function to compute the `cosh` and `sinh` series simultaneously. This improves performance by avoiding redundant computation of the exponential and its inverse.
4. Integrated `rs_cosh_sinh` into the `rs_cosh` and `rs_sinh` implementations within the constant expression code.
5. Updated rs_sin and rs_cos to utilize the shared `rs_cos_sin` function for efficiency.


### Release Note
<!-- BEGIN RELEASE NOTES -->
- polys
   - Added `rs_cosh_sinh` to calculate faster `cosh`,  `sinh` series together.
<!-- END RELEASE NOTES -->

Jatinbhardwaj-093 committed:
Typo fix

Jatinbhardwaj-093 committed:
Added rs_cosh_sinh function. Adjust the rs_sin,rs_cos to use rs_cos_sin. Adjust the rs_sinh,rs_cosh to use rs_cosh_sinh

Jatinbhardwaj-093 committed:
code quality

Jatinbhardwaj-093 committed:
change sin to sinh

oscarbenjamin commented on sympy/polys/ring_series.py:
```diff
@@ -1777,12 +1771,26 @@ def rs_cosh(p, x, prec):
     t1 = rs_series_inversion(t, x, prec)
     return (t + t1)/2
 
+def rs_cosh_sinh(p, x, prec):
+    r"""
+    Return the tuple ``(rs_cosh(p, x, prec)`, `rs_sinh(p, x, prec))``.
+
+    Is faster than calling rs_cosh and rs_sinh separately
```
Can you add this function to the docs in `doc/src/modules/polys/ringseries.rst` and also make a better docstring?

Jatinbhardwaj-093 replied:
>Can you add this function to the docs in doc/src/modules/polys/ringseries.rst and also make a better docstring?
Ok

Okay.

I have a question:

The functions `rs_cos_sin` and the new `rs_cosh_sinh` currently don’t handle the case where the expression is a constant.
I believe `rs_cos_sin` was originally designed to be used internally by other functions, which might be why constant handling wasn’t included.

Should I add constant handling directly in these functions, or should we leave them as-is.

oscarbenjamin replied:
I think leave them as is. If it is not difficult to do so then it would be good to at least make sure that they raise errors for invalid inputs.

Jatinbhardwaj-093 replied:
For rs_cos_sin its available. I added for the rs_cosh_sinh.

oscarbenjamin replied:
I don't see any tests with a nonzero constant term.

Jatinbhardwaj-093 committed:
Improve Docs

Jatinbhardwaj-093 committed:
Make rs_cos_sin 2.5x faster. In rs_cos_sin, rs_cosh_sinh added handling for the nonzero constant. Added repective test case.

Jatinbhardwaj-093 commented:
> as they first attempt to compute the series using functions that already account for constants.

Initially, I assumed the series was being computed using functions that already handle constant terms correctly. However, I take that back.

While it’s true that the series computation functions manage constant expressions, the main issue arose after the series was computed. Specifically, in domains like QQ, the underlying ring changes, which makes the original generator invalid. As a result, passing the generator to further computations—such as squaring or inversion—leads to errors because the generator is no longer in the ring.

I’ve now added proper handling for non-zero constants, along with an example to illustrate the fix.

While resolving this, I also checked the performance. The new rs_cosh_sinh works exactly as intended and is 2× faster than calling the respective functions separately.

However, rs_cos_sin performed worse—it was 2× slower than calling the individual functions. I added manual looping for smaller polynomials.

Updated performance results:

- Multivariate case and univariate polynomials with length < 20: ~1.1× faster
     This isn’t a huge improvement, but it doesn’t make things worse either.
- Univariate polynomials with length > 20: ~2.2× faster

If there’s any way to make the algorithm more efficient,  let me know!

Jatinbhardwaj-093 committed:
More test Case

Jatinbhardwaj-093 commented:
Previously, it was giving around 60% test coverage. To improve this, I added cases involving numerical constants, which raised the coverage to ~80% for the functions: `sin, cos, cos_sin, sinh, cosh, and cosh_sinh`.

I also added two new tests in Puiseux: one for `cosh` and another for the newly added function `rs_cosh_sin`. For Puiseux, coverage is now ~95%.

Currently, only edge cases like None inputs and very large polynomials are left untested. But overall, I think the existing cases now provide solid and comprehensive coverage.

oscarbenjamin commented:
Is the coverage 100% for `rs_sin_cos` and `rs_sinh_cosh`?

Jatinbhardwaj-093 committed:
fix the ordering of sin, cos in rs_cos_sin

Jatinbhardwaj-093 commented:
> Is the coverage 100% for `rs_sin_cos` and `rs_sinh_cosh`?

No it around 80%. Missing two cases where 

1. polynomial is none
2. Falling case of nonzero constant

oscarbenjamin commented:
> No it around 80%.

Can it be made 100%



> * polynomial is none
> * Falling case of nonzero constant

When would either of these things happen?

Jatinbhardwaj-093 commented:
> Can it be made 100% 

Yes

> When would either of these things happen?

1. The first case occurs when the input polynomial p is empty (i.e., the zero polynomial). This can be easily tested with a simple case. Below code block is at the very start of the function which early exit this case is left.
```python 
R = p.ring
if not p:
   return R(0) 
```

2. The second case is a safety mechanism to prevent operations that aren’t supported by the current domain of the polynomial ring. The `else` block for this scenario hasn’t been implemented yet:
```python
if c:
        if R.domain is EX:
            c_expr = c.as_expr()
            t1, t2 = sin(c_expr), cos(c_expr)
        elif isinstance(c, PolyElement):
            try:
                c_expr = c.as_expr()
                t1, t2 = R(sin(c_expr)), R(cos(c_expr))
            except ValueError:
                R = R.add_gens([sin(c_expr), cos(c_expr)])
                p = p.set_ring(R)
                x = x.set_ring(R)
                c = c.set_ring(R)
                t1, t2 = R(sin(c_expr)), R(cos(c_expr))
        else:
            try:
                t1, t2 = R(sin(c)), R(cos(c))
            except ValueError:
                raise DomainError("The given series cannot be expanded in "
                    "this domain.")
```
I’m currently adding a test case for this.

3. There’s also a third case in `rs_cos_sin` when the polynomial degree is greater than 20. In that case, the code opts to compute the series using the tangent expansion, which is faster than the manual loop.
---
For now, I’m adding the test case for the second scenario. Let me know if you’d like me to add tests for the first and third cases as well.

Jatinbhardwaj-093 commented:
I agree. I think the try-except block alone should be sufficient to handle all cases. I’ll verify it to ensure it doesn’t break any existing tests.

Jatinbhardwaj-093 committed:
refactor the constant block

Jatinbhardwaj-093 committed:
remove unnecessary import

Jatinbhardwaj-093 commented:
Some tests didn’t run , so I’ve closed and reopened the PR.

oscarbenjamin commented:
Looks good. Thanks

Jatinbhardwaj-093 commented:
Should I refactor other parts of the file?
I believe the `ring_series.py` section could benefit from a review and some refactoring in multiple areas.
Also, after running code coverage, I noticed that several functions have very low test coverage.
Would it be a good idea to focus on improving this?

oscarbenjamin commented:
It is generally good to improve test coverage. I typically find that bugs are discovered in the process if the code was not well tested before.

Ultimately I do think that all of this should be rewritten in a different way but it is good for the current code to be well tested and bug free.
